File: Code/Mantid/scripts/reduction_workflow/instruments/sans/hfir_instrument.py
#pylint: disable=invalid-name
import sys

def get_default_beam_center(workspace=None):
    """
        Returns the default beam center position, or the pixel location
        of real-space coordinates (0,0).
    """
    # If a workspace is not provided, we'll figure out the position later
    if workspace is None:
        return [None, None]
    return get_pixel_from_coordinate(0, 0, workspace)

def get_pixel_from_coordinate(x, y, workspace):
    """
        Returns the pixel coordinates corresponding to the
        given real-space position.

        This assumes that the center of the detector is aligned
        with the beam. An additional offset may need to be applied

        @param x: real-space x coordinate [m]
        @param y: real-space y coordinate [m]
        @param workspace: the pixel number and size info will be taken from the workspace
    """
    nx_pixels, ny_pixels, pixel_size_x, pixel_size_y = _get_pixel_info(workspace)

    return [x/pixel_size_x*1000.0 + nx_pixels/2.0-0.5,
            y/pixel_size_y*1000.0 + ny_pixels/2.0-0.5]

def get_coordinate_from_pixel(x, y, workspace):
    """
        Returns the real-space coordinates corresponding to the
        given pixel coordinates [m].

        This assumes that the center of the detector is aligned
        with the beam. An additional offset may need to be applied

        @param x: pixel x coordinate
        @param y: pixel y coordinate
        @param workspace: the pixel number and size info will be taken from the workspace
    """
    nx_pixels, ny_pixels, pixel_size_x, pixel_size_y = _get_pixel_info(workspace)

    return [(x-nx_pixels/2.0+0.5) * pixel_size_x/1000.0,
            (y-ny_pixels/2.0+0.5) * pixel_size_y/1000.0]

def get_masked_pixels(nx_low, nx_high, ny_low, ny_high, workspace):
    """
        Generate a list of masked pixels.
        @param nx_low: number of pixels to mask on the lower-x side of the detector
        @param nx_high: number of pixels to mask on the higher-x side of the detector
        @param ny_low: number of pixels to mask on the lower-y side of the detector
        @param ny_high: number of pixels to mask on the higher-y side of the detector
        @param workspace: the pixel number and size info will be taken from the workspace
    """
    nx_pixels, ny_pixels, dummy_pixel_size_x, dummy_pixel_size_y = _get_pixel_info(workspace)
    if nx_low<0 or nx_high<0 or ny_low<0 or ny_high<0:
        raise RuntimeError, "Pixel edges should be greater than zero"

    masked_x = range(0, nx_low)
    masked_x.extend(range(nx_pixels-nx_high, nx_pixels))

    masked_y = range(0, ny_low)
    masked_y.extend(range(ny_pixels-ny_high, ny_pixels))

    masked_pts = []
    for y in masked_y:
        masked_pts.extend([ [y,x] for x in range(nx_pixels) ])
    for x in masked_x:
        masked_pts.extend([ [y,x] for y in range(ny_low, ny_pixels-ny_high) ])

    return masked_pts

def _get_pixel_info(workspace):
    """
        Get the pixel size and number of pixels from the workspace
        @param workspace: workspace to extract the pixel information from
    """
    ## Number of detector pixels in X
    nx_pixels = int(workspace.getInstrument().getNumberParameter("number-of-x-pixels")[0])
    ## Number of detector pixels in Y
    ny_pixels = int(workspace.getInstrument().getNumberParameter("number-of-y-pixels")[0])
    ## Pixel size in mm
    pixel_size_x = workspace.getInstrument().getNumberParameter("x-pixel-size")[0]
    pixel_size_y = workspace.getInstrument().getNumberParameter("y-pixel-size")[0]

    return nx_pixels, ny_pixels, pixel_size_x, pixel_size_y

def get_detector_from_pixel(pixel_list):
    """
        Returns a list of detector IDs from a list of [x,y] pixels,
        where the pixel coordinates are in pixel units.
    """
    return [ 1000000 + p[0]*1000 + p[1] for p in pixel_list ]

def get_aperture_distance(workspace):
    """
        Return the aperture distance
        @param workspace: workspace to get the aperture distance from
    """
    try:
        nguides = workspace.getRun().getProperty("number-of-guides").value
        apertures_lst = workspace.getInstrument().getStringParameter("aperture-distances")[0]
        apertures = apertures_lst.split(',')
        # Note that they are in reverse order, the first item is for 8 guides
        # and the last item is for 0 guide.
        index = 8-nguides
        return float(apertures[index])
    except:
        raise RuntimeError, "Could not find the for %s\n  %s" % (workspace, sys.exc_value)

File: Framework/WorkflowAlgorithms/src/HFIRInstrument.cpp
// Mantid Repository : https://github.com/mantidproject/mantid
//
// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
//   NScD Oak Ridge National Laboratory, European Spallation Source,
//   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
// SPDX - License - Identifier: GPL - 3.0 +
#include "MantidWorkflowAlgorithms/HFIRInstrument.h"
#include "MantidAPI/Run.h"
#include "MantidDataObjects/Workspace2D.h"
#include "MantidGeometry/Instrument.h"
#include "MantidKernel/DataService.h"
#include "MantidKernel/Exception.h"
#include "MantidKernel/Logger.h"
#include "MantidKernel/Property.h"
#include "MantidKernel/PropertyWithValue.h"
#include "MantidKernel/StringTokenizer.h"
#include "Poco/NumberParser.h"

#include <memory>
#include <utility>

#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/regex.hpp>

namespace Mantid::WorkflowAlgorithms::HFIRInstrument {
/**
    File change history is stored at: <https://github.com/mantidproject/mantid>
    Code Documentation is available at: <http://doxygen.mantidproject.org>
*/

/*
 * Read a parameter from the instrument description
 */
double readInstrumentParameter(const std::string &parameter, const API::MatrixWorkspace_sptr &dataWS) {
  std::vector<double> pars = dataWS->getInstrument()->getNumberParameter(parameter);
  if (pars.empty())
    throw Kernel::Exception::InstrumentDefinitionError("Unable to find [" + parameter + "] instrument parameter");
  return pars[0];
}

/*
 * Return the detector ID corresponding to the [x,y] pixel coordinates.
 */
int getDetectorFromPixel(const int &pixel_x, const int &pixel_y, const API::MatrixWorkspace_sptr &dataWS) {
  UNUSED_ARG(dataWS);
  return 1000000 + 1000 * pixel_x + pixel_y;
}

/*
 * Returns the real-space coordinates corresponding to the
 * given pixel coordinates [m].
 */
void getCoordinateFromPixel(const double &pixel_x, const double &pixel_y, const API::MatrixWorkspace_sptr &dataWS,
                            double &x, double &y) {
  const int nx_pixels = static_cast<int>(readInstrumentParameter("number-of-x-pixels", dataWS));
  const int ny_pixels = static_cast<int>(readInstrumentParameter("number-of-y-pixels", dataWS));
  const double pixel_size_x = readInstrumentParameter("x-pixel-size", dataWS);
  const double pixel_size_y = readInstrumentParameter("y-pixel-size", dataWS);
  x = (pixel_x - nx_pixels / 2.0 + 0.5) * pixel_size_x / 1000.0;
  y = (pixel_y - ny_pixels / 2.0 + 0.5) * pixel_size_y / 1000.0;
}

/*
 * Returns the pixel coordinates corresponding to the given real-space position.
 * This assumes that the center of the detector is aligned
 * with the beam. An additional offset may need to be applied
 * @param x: real-space x coordinate [m]
 * @param y: real-space y coordinate [m]
 *
 */
void getPixelFromCoordinate(const double &x, const double &y, const API::MatrixWorkspace_sptr &dataWS, double &pixel_x,
                            double &pixel_y) {
  const int nx_pixels = static_cast<int>(readInstrumentParameter("number-of-x-pixels", dataWS));
  const int ny_pixels = static_cast<int>(readInstrumentParameter("number-of-y-pixels", dataWS));
  const double pixel_size_x = readInstrumentParameter("x-pixel-size", dataWS);
  const double pixel_size_y = readInstrumentParameter("y-pixel-size", dataWS);
  pixel_x = x / pixel_size_x * 1000.0 + nx_pixels / 2.0 - 0.5;
  pixel_y = y / pixel_size_y * 1000.0 + ny_pixels / 2.0 - 0.5;
}

/*
 * Returns the default beam center position, or the pixel location
 * of real-space coordinates (0,0).
 */
void getDefaultBeamCenter(const API::MatrixWorkspace_sptr &dataWS, double &pixel_x, double &pixel_y) {
  getPixelFromCoordinate(0.0, 0.0, dataWS, pixel_x, pixel_y);
}

/*
 * Get the source to sample distance (ssd)
 * If "Header/source_distance exists", ssd is this
 * otherwise get the guides distance (based on the number of guides used),
 * defined in instrument parameters file as "aperture-distances"
 * and sums "Header/sample_aperture_to_flange"
 */
double getSourceToSampleDistance(const API::MatrixWorkspace_sptr &dataWS) {

  double sourceToSampleDistance;

  std::vector<double> parsDouble = dataWS->getInstrument()->getNumberParameter("Header_source_distance");
  if (!parsDouble.empty()) {
    // First let's try to get source_distance first:
    sourceToSampleDistance = parsDouble[0] * 1000; // convert to mm
  } else {
    const auto nGuides = dataWS->run().getPropertyValueAsType<int>("Motor_Positions_nguides");
    // aperture-distances: array from the instrument parameters
    std::vector<std::string> parsString = dataWS->getInstrument()->getStringParameter("aperture-distances");
    if (parsString.empty())
      throw Kernel::Exception::InstrumentDefinitionError("Unable to find [aperture-distances] instrument parameter");
    std::string guidesDistances = parsString[0];

    std::vector<std::string> guidesDistancesSplit;
    boost::split(guidesDistancesSplit, guidesDistances, boost::is_any_of("\t ,"), boost::token_compress_on);
    sourceToSampleDistance = boost::lexical_cast<double>(guidesDistancesSplit[nGuides]);

    auto sourceToSampleDistanceOffset =
        dataWS->run().getPropertyValueAsType<double>("Header_sample_aperture_to_flange");

    sourceToSampleDistance -= sourceToSampleDistanceOffset;
  }
  return sourceToSampleDistance;
}

} // namespace Mantid::WorkflowAlgorithms::HFIRInstrument
